src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs

```diff
@@ -102,5 +119,62 @@ public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocumen
             {
             }
         }
+
+        private readonly struct FirstStatementTracker
+        {
+            private readonly HashSet<StatementSyntax> _removed;
+            public readonly HashSet<StatementSyntax> FirstStatements;
+
+            public FirstStatementTracker(bool unused)
+            {
+                _removed = new HashSet<StatementSyntax>();
+                FirstStatements = new HashSet<StatementSyntax>();
+            }
+
+            private bool WasFirst(StatementSyntax statementToRemove)
+            {
+                if (statementToRemove.IsFirstStatementInEnclosingBlock() || statementToRemove.IsFirstStatementInSwitchSection())
+                {
+                    return true;
+                }
+
+                bool wasMadeFirstByRemovalOfPredecessor = FirstStatements.Remove(statementToRemove);
+                if (wasMadeFirstByRemovalOfPredecessor)
+                {
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryFindFirstUnremovedSuccessor(StatementSyntax statement, out StatementSyntax result)
+            {
+                result = statement;
+                do
+                {
+                    result = result.GetNextStatement();
+                } while (_removed.Contains(result));
+
+                return result != null;
+            }
+
+            private void MakeFirstUnremovedSuccessorFirst(StatementSyntax statement)
+            {
+                if (TryFindFirstUnremovedSuccessor(statement, out var firstUnremovedSuccessor))
```

Comment:
Rather than trying to track what the new next statement is, why not just track the containing block? i.e. just keep a `HashSet<BlockSyntax>`, and in your cleanup step just loop through those and re-grab the first statement? I suspect it'll be easier than trying to maintain the logic here.

Reply:
And I suspect this might make it a lot simpler to follow: I admit this helper class with the different methods is pretty tricky to review.